src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -0,0 +1,253 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "network-commissioning.h"
+
+#include <cstring>
+
+#include <lib/support/CodeUtils.h>
+#include <lib/support/ReturnMacros.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConnectivityManager.h>
+#include <transport/NetworkProvisioning.h>
+
+#if defined(CHIP_DEVICE_LAYER_TARGET)
+#define DEVICENETWORKPROVISIONING_HEADER <platform/CHIP_DEVICE_LAYER_TARGET/DeviceNetworkProvisioningDelegateImpl.h>
+#include DEVICENETWORKPROVISIONING_HEADER
+#endif
+
+namespace chip {
+namespace app {
+namespace clusters {
+namespace NetworkCommissioning {
+
+constexpr uint8_t kMaxNetworkIDLen       = 32;
+constexpr uint8_t kMaxThreadDatasetLen   = 128;
+constexpr uint8_t kMaxWiFiSSIDLen        = 32;
+constexpr uint8_t kMaxWiFiCredentialsLen = 64;
+constexpr uint8_t kMaxNetworks           = 4;
+constexpr char kThreadNetworkName[]      = "tmp-thread-network";
+
+// TODO: The network info should be stored in attribute storage instead of memory.
+// TODO: We cannot get length of operational dataset here, should be fixed after we have byte string type.
+// TODO: Should encode response for command responses.
+enum class NetworkCommissioninError : uint8_t
+{
+    kSuccess                = 0,
+    kOutOfRange             = 1,
+    kBoundsExceeded         = 2,
+    kNetworkIDNotFound      = 3,
+    kDuplicateNetworkID     = 4,
+    kNetworkNotFound        = 5,
+    kRegulatoryError        = 6,
+    kAuthFailure            = 7,
+    kUnsupportedSecurity    = 8,
+    kOtherConnectionFailure = 9,
+    kIPV6Failed             = 10,
+    kIPBindFailed           = 11,
+    kLabel9                 = 12,
```

Comment:
do we need kLabel here? What do they mean?

Reply:
This is defined in the command definition, I realized that these enums should be generated by code gen so I removed them.